File: bpui/parse_blocks.py
"""Strict codeblock parser for blueprint outputs."""

import re
from typing import Optional, Dict, List
from .profiler import profile


# Ordered asset names as they appear in output
ASSET_ORDER = [
    "system_prompt",
    "post_history",
    "character_sheet",
    "intro_scene",
    "intro_page",
    "a1111",
    "suno",
]

# Filename mapping
ASSET_FILENAMES = {
    "system_prompt": "system_prompt.txt",
    "post_history": "post_history.txt",
    "character_sheet": "character_sheet.txt",
    "intro_scene": "intro_scene.txt",
    "intro_page": "intro_page.md",
    "a1111": "a1111_prompt.txt",
    "suno": "suno_prompt.txt",
}


class ParseError(Exception):
    """Parse error."""
    pass


def extract_codeblocks(text: str) -> List[str]:
    """Extract all fenced codeblocks from text.

    Args:
        text: LLM output

    Returns:
        List of codeblock contents (without fence markers)
    """
    # Match ```...``` blocks (handles language tags)
    pattern = r"```(?:[a-z]*\n)?(.*?)```"
    matches = re.findall(pattern, text, re.DOTALL)
    return [m.strip() for m in matches]


def parse_blueprint_output(text: str) -> Dict[str, str]:
    """Parse blueprint orchestrator output into assets.

    Args:
        text: LLM output containing codeblocks

    Returns:
        Dict mapping asset names to content

    Raises:
        ParseError: If output doesn't match expected structure
    """
    with profile("parse_blueprint_output", input_length=len(text)):
        blocks = extract_codeblocks(text)

        if len(blocks) == 0:
            raise ParseError("No codeblocks found in output")

        # Check for optional Adjustment Note
        start_idx = 0
        adjustment_note = None

        if blocks[0].strip().startswith("Adjustment Note:"):
            adjustment_note = blocks[0].strip()
            start_idx = 1

    # Require exactly 7 asset blocks after adjustment note (if present)
    asset_blocks = blocks[start_idx:]
    if len(asset_blocks) != 7:
        raise ParseError(
            f"Expected 7 asset blocks, found {len(asset_blocks)}. "
            f"Required order: {', '.join(ASSET_ORDER)}"
        )

        # Map blocks to asset names
        assets = {}
        for i, asset_name in enumerate(ASSET_ORDER):
            assets[asset_name] = asset_blocks[i]

        return assets


def extract_single_asset(output: str, asset_name: str) -> str:
    """Extract a single asset from LLM output.

    Args:
        output: Raw LLM output (may contain adjustment note + codeblock)
        asset_name: Name of the asset to extract

    Returns:
        Asset content (text inside the codeblock)

    Raises:
        ParseError: If parsing fails
    """
    blocks = extract_codeblocks(output)

    if not blocks:
        raise ParseError(f"No codeblocks found for {asset_name}")

    # Check for adjustment note
    start_idx = 0
    if blocks[0].strip().startswith("Adjustment Note:"):
        start_idx = 1

    asset_blocks = blocks[start_idx:]
    if not asset_blocks:
        raise ParseError(f"No asset block found for {asset_name}")

    # Return the first (and should be only) asset block
    return asset_blocks[0]


def extract_character_name(character_sheet: str) -> Optional[str]:
    """Extract character name from character sheet.

    Args:
        character_sheet: Character sheet content

    Returns:
        Sanitized character name or None
    """
    # Look for "name: ..." line
    match = re.search(r"^name:\s*(.+)$", character_sheet, re.MULTILINE | re.IGNORECASE)
    if not match:
        return None

    name = match.group(1).strip()
    # Sanitize using same logic as export script
    sanitized = re.sub(r"[^a-z0-9]+", "_", name.lower())
    sanitized = re.sub(r"_+", "_", sanitized)
    sanitized = sanitized.strip("_")
    return sanitized

File: packages/shared/src/parse/parse-blocks.ts
/**
 * Codeblock parser for blueprint outputs.
 * Extracts and validates assets from LLM-generated content.
 */

import type {
  AssetDefinition as TypesAssetDefinition,
  Template as TypesTemplate,
} from '../types';

// Re-export types from types/index with local type names to avoid conflicts
export type AssetDefinition = TypesAssetDefinition;
export type Template = TypesTemplate;

// Ordered asset names as they appear in output
export const ASSET_ORDER: ReadonlyArray<AssetName> = [
  'system_prompt',
  'post_history',
  'character_sheet',
  'intro_scene',
  'intro_page',
  'a1111',
  'suno',
] as const;

export type AssetName = typeof ASSET_ORDER[number];

export interface ParseResult {
  assets: Record<string, string>;
  adjustmentNote?: string;
}

export class ParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ParseError';
  }
}

// Asset order for parsing
const DEFAULT_ASSET_ORDER: AssetName[] = [
  'system_prompt',
  'post_history',
  'character_sheet',
  'intro_scene',
  'intro_page',
  'a1111',
];

// Default filename mapping (can be overridden by templates)
export const DEFAULT_ASSET_FILENAMES: Record<AssetName, string> = {
  system_prompt: 'system_prompt.txt',
  post_history: 'post_history.txt',
  character_sheet: 'character_sheet.txt',
  intro_scene: 'intro_scene.txt',
  intro_page: 'intro_page.md',
  a1111: 'a1111_prompt.txt',
  suno: 'suno_prompt.txt',
};

/**
 * Extract all fenced codeblocks from text.
 * Does NOT handle nested codeblocks (acceptable for intended use case).
 */
export function extractCodeblocks(text: string): string[] {
  // Match ```...``` blocks (handles language tags)
  const pattern = /```(?:[a-z]*\n)?(.*?)```/gs;
  const matches = text.match(pattern);
  if (!matches) return [];
  return matches.map(m => m.trim());
}

/**
 * Parse blueprint orchestrator output into assets.
 * @param text - LLM output containing codeblocks
 * @param template - Optional template defining expected assets
 * @returns Dict mapping asset names to content
 * @throws ParseError if output doesn't match expected structure
 */
export function parseBlueprintOutput(text: string, template?: TypesTemplate): ParseResult {
  const blocks = extractCodeblocks(text);

  if (blocks.length === 0) {
    throw new ParseError('No codeblocks found in output');
  }

  // Check for optional Adjustment Note
  let startIdx = 0;
  let adjustmentNote: string | undefined;

  if (blocks[0].trim().startsWith('Adjustment Note:')) {
    adjustmentNote = blocks[0].trim();
    startIdx = 1;
  }

  const assetBlocks = blocks.slice(startIdx);

  // Determine expected asset order from template or use default
  let expectedAssets: AssetName[] = [];
  if (template && template.assets.length > 0) {
    expectedAssets = template.assets.map(a => a.name as AssetName);
  } else {
    expectedAssets = DEFAULT_ASSET_ORDER;
  }

  const expectedCount = expectedAssets.length;

  // Validate asset count
  if (assetBlocks.length !== expectedCount) {
    const blockPreviews = assetBlocks
      .slice(0, 3)
      .map((b, i) => `  Block ${i}: ${b.substring(0, 75)}${b.length > 75 ? '...' : ''}`)
      .join('\n');

    let errorMessage = `Expected ${expectedCount} asset blocks, found ${assetBlocks.length}. `;
    errorMessage += `Template requires order: ${expectedAssets.join(', ')}\n`;
    errorMessage += `Actual blocks found:\n${blockPreviews}`;

    if (assetBlocks.length > 3) {
      errorMessage += `\n  ... and ${assetBlocks.length - 3} more blocks`;
    }

    throw new ParseError(errorMessage);
  }

  // Map blocks to asset names
  const assets: Record<string, string> = {};
  for (let i = 0; i < expectedAssets.length; i++) {
    assets[expectedAssets[i]] = assetBlocks[i];
  }

  // Validate generated content for fatal contract violations
  const contentFailures = validateAssetsContent(assets);
  if (contentFailures && Object.keys(contentFailures).length > 0) {
    const details = Object.entries(contentFailures)
      .map(([asset, issues]) => `${asset}: ${Array.from(new Set(issues)).join(', ')}`)
      .join('; ');
    throw new ParseError('Generated content failed validation checks: ' + details);
  }

  return { assets, adjustmentNote };
}

/**
 * Extract a single asset from LLM output.
 */
export function extractSingleAsset(output: string, assetName: AssetName): string {
  const blocks = extractCodeblocks(output);

  if (!blocks.length) {
    throw new ParseError(`No codeblocks found for ${assetName}`);
  }

  // Check for adjustment note
  let startIdx = 0;
  if (blocks[0].trim().startsWith('Adjustment Note:')) {
    startIdx = 1;
  }

  const assetBlocks = blocks.slice(startIdx);
  if (!assetBlocks.length) {
    throw new ParseError(`No asset block found for ${assetName}`);
  }

  return assetBlocks[0];
}

/**
 * Extract character name from character sheet.
 */
export function extractCharacterName(characterSheet: string): string | null {
  const displayName = extractCharacterDisplayName(characterSheet);
  if (!displayName) return null;
  return sanitizeCharacterName(displayName);
}

/**
 * Extract a human-readable character name from asset content.
 */
export function extractCharacterDisplayName(content: string): string | null {
  const match = content.match(/^name:\s*(.+)$/m);
  if (!match) return null;

  const name = match[1].trim().replace(/^['"]|['"]$/g, '');
  return name || null;
}

/**
 * Sanitize a character name for filesystem-safe identifiers.
 */
export function sanitizeCharacterName(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/_+/g, '_')
    .replace(/^_|_$/g, '');
}

/**
 * Infer a character's display name from generated asset contents.
 */
export function inferCharacterDisplayNameFromAssets(
  assets: Record<string, string>,
  preferredAssets: Iterable<string> = ['character_sheet', 'char_basic_info']
): string | null {
  const orderedAssetNames: AssetName[] = [];
  const seen = new Set<AssetName>();

  for (const asset of preferredAssets) {
    if (asset in assets && !seen.has(asset)) {
      orderedAssetNames.push(asset);
      seen.add(asset);
    }
  }

  for (const asset of Object.keys(assets)) {
    if (!seen.has(asset)) {
      orderedAssetNames.push(asset);
      seen.add(asset);
    }
  }

  for (const assetName of orderedAssetNames) {
    const displayName = extractCharacterDisplayName(assets[assetName] || '');
    if (displayName) return displayName;
  }

  return null;
}

/**
 * Infer a sanitized filesystem-safe character name from asset contents.
 */
export function inferCharacterNameFromAssets(
  assets: Record<string, string>,
  preferredAssets?: Iterable<string>
): string | null {
  const displayName = inferCharacterDisplayNameFromAssets(assets, preferredAssets);
  if (!displayName) return null;
  return sanitizeCharacterName(displayName);
}

// ============================================================================
// Content Validation
// ============================================================================

const PLACEHOLDER_PATTERNS: ReadonlyArray<[string, RegExp]> = [
  ['{PLACEHOLDER}', /\{PLACEHOLDER\}/g],
  ['Suno {TITLE}', /\{TITLE\}/g],
  [
    'Suno other {..}',
    /\{GENRE\}|\{STYLE\}|\{MOOD\}|\{ENERGY\}|\{TEMPO\}|\{BPM\}|\{TEXTURE\}|\{Remaster Style\}/g,
  ],
  ['A1111 slot ((...))', /\(\(\.\.\)\)/g],
  [
    'A1111 any slot ((...something...)) left',
    /\(\([^\)]*\.\.\.[^\)]*\)/g,
  ],
  [
    'Character sheet bracket placeholders',
    /\[(?!\")["'][A-Z][^\]]*\]/g,
  ],
] as const;

const USER_AUTHORITY_PATTERNS: ReadonlyArray<[string, RegExp]> = [
  [
    'Narrates {{user}} action/thought/consent',
    /\{\{user\}\}\s+(?:is|was|feels?|felt|thinks?|thought|decides?|decided|knows?|knows?|wants?|wanted|says?|said|nods?|smiles?|walks?|steps?|looks?|touches?|takes?|gives?|allows?|consents?|agrees?|gasps?|moans?)\b/gi,
  ],
  [
    'Narrates {{user}} internal state',
    /\{\{user\}\}'s\s+(?:mind|thoughts?|feelings?|emotions?|desire|consent|decision|reaction|actions?)\b/gi,
  ],
] as const;

const INSTRUCTIONAL_USER_REFERENCE_PATTERN = /(?:never|do not|don't|avoid|without)\s+(?:narrat(?:e|ing)|describ(?:e|ing)|assign(?:ing)?)\s*$/gi;

const CHARACTER_SHEET_FILENAME = 'character_sheet.txt';

/**
 * Return placeholder-related validation issue labels for a file's content.
 */
function detectPlaceholderIssues(text: string, filename: string): string[] {
  const issues: string[] = [];

  for (const [label, pattern] of PLACEHOLDER_PATTERNS) {
    if (label === 'Character sheet bracket placeholders' && filename !== CHARACTER_SHEET_FILENAME) {
      continue;
    }
    if (pattern.test(text)) {
      issues.push(label);
    }
  }

  return issues;
}

/**
 * Return user-authorship issue labels for content that narrates {{user}}.
 */
function detectUserAuthorityIssues(text: string): string[] {
  const issues: string[] = [];

  for (const [label, pattern] of USER_AUTHORITY_PATTERNS) {
    for (const match of text.matchAll(pattern)) {
      const contextBefore = text.substring(Math.max(0, match.index! - 48), match.index!).trim();
      if (!INSTRUCTIONAL_USER_REFERENCE_PATTERN.test(contextBefore)) {
        issues.push(label);
        break;
      }
    }
  }

  return issues;
}

/**
 * Validate a generated asset's content and return issue labels.
 */
export function validateAssetContent(assetName: AssetName, content: string): string[] {
  let filename = `${assetName}.txt`;
  if (assetName === 'intro_page') filename = 'intro_page.md';
  if (assetName === 'character_sheet') filename = CHARACTER_SHEET_FILENAME;

  const issues = detectPlaceholderIssues(content, filename);
  issues.push(...detectUserAuthorityIssues(content));
  return issues;
}

/**
 * Validate all generated assets and return {asset_name: issues} for failures.
 */
export function validateAssetsContent(assets: Record<string, string>): Record<string, string[]> | null {
  const failures: Record<string, string[]> = {};
  for (const [assetName, content] of Object.entries(assets)) {
    const issues = validateAssetContent(assetName as AssetName, content);
    if (issues.length > 0) {
      failures[assetName] = issues;
    }
  }
  return Object.keys(failures).length > 0 ? failures : null;
}
